kafka/consumer.go

```diff
@@ -494,6 +494,10 @@ func (c *Consumer) ReadMessage(timeout time.Duration) (*Message, error) {
 // Close Consumer instance.
 // The object is no longer usable after this call.
 func (c *Consumer) Close() (err error) {
+	if c.IsClosed() {
+		return getOperationNotAllowedErrorForClosedClient()
+	}
+
 	var newValue uint32 = 1
 	currentValue := atomic.LoadUint32(&c.isClosed)
 	if !atomic.CompareAndSwapUint32(&c.isClosed, currentValue, newValue) {
```

Comment:
You should add `currentValue == 1 || ` here and remove isClosed before because if two threads reach line 501, then one of them reaches line 508 and the second executes line 502 and 503 will execute a compare and swap 1 -> 1 that returns true (the value hasn't changed since when it was read) and continues so it's closing two times

Reply:
does it still solve our problem? 
because  `currentValue == 1 || !atomic.CompareAndSwapUint32(&c.isClosed, currentValue, newValue)` is still not atomic. Can we instead introduce a mutex to perform this entirely in atomic fashion ?

Reply:
Why it's not atomic? If it's already 1 it doesn't need to try to set to 1.

To further simplify, you can replace with:

```suggestion
	if !atomic.CompareAndSwapUint32(&c.isClosed, 0, 1) {
```

without anything else, it's the only case where it has to succeed.

Reply:
got it.